File: pigeons/printer_api.dart
import 'package:pigeon/pigeon.dart';

/// {@nodoc}
class ConnectionInfo {
  const ConnectionInfo({
    this.ipAddress,
    this.macAddress,
    this.port,
  });

  final String? ipAddress;
  final String? macAddress;
  final int? port;
}

/// Enumeration of the various printer control languages supported by Zebra Printers.
/// {@category Printer}
enum PrinterLanguage {
  /// Printer control language ZPL.
  zpl,

  /// Printer control language CPCL.
  cpcl,

  /// Printer control language line_print mode.
  linePrint;
}

/// Describe printer status.
/// {@category Printer}
@SwiftFunction('ZPrinterStatus')
enum PrinterStatus {
  /// The head is cold.
  headCold,

  /// The head is open.
  headOpen,

  /// The head is too hot.
  headTooHot,

  /// The paper is out.
  paperOut,

  /// There is a partial format in progress.
  partialFormatInProgress,

  /// The printer is paused.
  paused,

  /// The printer reports back that it is ready to print.
  readyToPrint,

  /// The receive buffer is full.
  receiveBufferFull,

  /// The ribbon is out.
  ribbonOut;
}

@HostApi()

/// {@nodoc}
abstract class PrinterAPI {
  /// Returns a [PrinterStatus] that can be used to determine the status of a printer.
  PrinterStatus currentStatus(ConnectionInfo info);

  /// Returns the printer control language (e.g. ZPL or CPCL) of the printer.
  PrinterLanguage controlLanguage(ConnectionInfo info);

  /// Prints an image to the connected device as a monochrome image.
  void printImage(ConnectionInfo info, Uint8List data);

  /// Sends the appropriate calibrate command to the printer.
  void calibrate(ConnectionInfo info);

  /// Sends the appropriate print configuration command to the printer.
  void printConfigurationLabel(ConnectionInfo info);

  /// Sends the appropriate restore defaults command to the printer.
  void restoreDefaults(ConnectionInfo info);

  /// Converts the specified command to bytes using the Java default charset and sends the bytes to the printer.
  void sendCommand(ConnectionInfo info, String command);

  /// Sends the appropriate reset command to the printer.
  void reset(ConnectionInfo info);
}

File: ios/Classes/Pigeons/PrinterAPI.swift
import Flutter

typealias ZConnection = NSObject & ZebraPrinterConnection

enum PrinterAPIError: Error {
    case notImplemented
    case notAssigned
}

final class ImplPrinterAPI: PrinterAPI {
    func currentStatus(info: ConnectionInfo) throws -> PrinterStatus {
        var result: PrinterStatus?
        let thr = ThreadExecutor(task: {
            do {
                let conn = try self.generalConnection(info: info)
                conn.open()
                
                defer {
                    conn.close()
                }
                
                let printer = try ZebraPrinterFactory.getInstance(conn)
                
                let status = try printer.getCurrentStatus()
                
                if status.isHeadCold {
                    result = PrinterStatus.headCold
                } else if status.isHeadOpen {
                    result = PrinterStatus.headOpen
                } else if status.isHeadTooHot {
                    result = PrinterStatus.headTooHot
                } else if status.isPaperOut {
                    result = PrinterStatus.paperOut
                } else if status.isPartialFormatInProgress {
                    result = PrinterStatus.partialFormatInProgress
                } else if status.isPaused {
                    result = PrinterStatus.paused
                } else if status.isReadyToPrint {
                    result = PrinterStatus.readyToPrint
                } else if status.isReceiveBufferFull {
                    result = PrinterStatus.receiveBufferFull
                } else if status.isRibbonOut {
                    result = PrinterStatus.ribbonOut
                }
            } catch {
                
            }
        })
        thr.start()
        thr.join()
        
        if result == nil {
            throw PrinterAPIError.notAssigned
        }
        
        return result!
    }
    
    func controlLanguage(info: ConnectionInfo) throws -> PrinterLanguage {
        return PrinterLanguage.zpl
    }
    
    func printImage(info: ConnectionInfo, data: FlutterStandardTypedData) throws {
        
    }
    
    func calibrate(info: ConnectionInfo) throws {
        
    }
    
    func printConfigurationLabel(info: ConnectionInfo) throws {
        
    }
    
    func restoreDefaults(info: ConnectionInfo) throws {
    
    }
    
    func sendCommand(info: ConnectionInfo, command: String) throws {
        
    }
    
    func reset(info: ConnectionInfo) throws {
        
    }
    
    
    private func generalConnection(info: ConnectionInfo) throws -> ZConnection {
        if let ipAddress = info.ipAddress {
            let portOrDefault = info.port != nil ? Int(info.port!) : 9100
            return TcpPrinterConnection(address: ipAddress, andWithPort: portOrDefault)
        }
        
        if let macAddress = info.macAddress {
            return MfiBtPrinterConnection(serialNumber: macAddress)
        }
        
        throw PrinterAPIError.notImplemented
    }
}
